plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/SpringReload.java

```diff
@@ -0,0 +1,572 @@
+        } else {
+            LOGGER.debug("try to add xml '{}' into {}", xml, ObjectUtils.identityToString(beanFactory));
+        }
+    }
+
+    public void appendAll(SpringReload content) {
+        classes.addAll(content.classes);
+        properties.addAll(content.properties);
+        xmls.addAll(content.xmls);
+        newScanBeanDefinitions.addAll(content.newScanBeanDefinitions);
+    }
+
+    public boolean reload() throws IOException {
+        boolean allowBeanDefinitionOverriding = BeanFactoryProcessor.isAllowBeanDefinitionOverriding(beanFactory);
+        long now = System.currentTimeMillis();
+        try {
+            beanFactoryAssistant.setReload(true);
+            LOGGER.info("****************************************************************************************************");
+            LOGGER.info("##### start reloading '{}'", ObjectUtils.identityToString(beanFactory));
+            LOGGER.trace("SpringReload:{},  beanFactory:{}", this, beanFactory);
+            BeanFactoryProcessor.setAllowBeanDefinitionOverriding(beanFactory, true);
+            return doReload();
+        } finally {
+            beanFactoryAssistant.increaseReloadTimes();
+            BeanFactoryProcessor.setAllowBeanDefinitionOverriding(beanFactory, allowBeanDefinitionOverriding);
+            LOGGER.info("##### [{}th] finish reloading '{}', it cost {}ms", beanFactoryAssistant.getReloadTimes(),
+                    ObjectUtils.identityToString(beanFactory), System.currentTimeMillis() - now);
+            LOGGER.info("****************************************************************************************************");
+        }
+    }
+
+    private boolean doReload() {
+        // 0. properties reload
+        if (!properties.isEmpty()) {
+            PropertyReload.reloadPropertySource(beanFactory);
+            recreatedBeans.addAll(PropertyReload.checkAndGetReloadBeanNames(beanFactory));
+            recreatedBeans.addAll(beanFactoryAssistant.placeHolderXmlRelation.keySet());
+        }
+        // 1. clear cache
+        clearCache();
+
+        // 2. reload xmls: the beans will be destroyed
+        Set<String> reloadDefinitionNames = XmlReload.reloadXmlsAndGetBean(beanFactory, !this.properties.isEmpty(),
+                beanFactoryAssistant.placeHolderXmlRelation, recreatedBeans, xmls);
+        destroyBeans.addAll(reloadDefinitionNames);
+        // 3. add changed class into recreate beans
+        List<String> needReloadBeanDefinitionNames = new ArrayList<>();
+        for (Class clazz : classes) {
+            destroyClasses.add(ClassUtils.getUserClass(clazz).getName());
+            String[] names = beanFactory.getBeanNamesForType(clazz);
+            if (names != null && names.length > 0) {
+                recreatedBeans.addAll(Arrays.asList(names));
+                // 3.1 when the bean is @Configuration, it should be recreated.
+                reloadAnnotatedBeanDefinitions(clazz, names, needReloadBeanDefinitionNames);
```

Comment:
'reloadannotationBeanDefinition' should be put in step 3.2. It is confusing to process configuration class in the same loop.